# Welcome to DataSpell!
DataSpell is an Integrated Development Environment (IDE) that is,
dedicated to specific tasks for exploratory data analysis and prototyping ML (machine learning) models.

In [1]:
# Dependencies and Setup
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import scipy.stats as st

We're going to be importing 2 large CSV files.
charts.csv - Contains the attributes:
track_id, name, country (country code), date (date appeared in countries' charts), position (position this track appeared in countries' charts), streams (number of streams the track had until 'date'), artists (all involved), artist_genres, duration (milliseconds), explicit (true/false)

dataset.csv - Contains the attributes:
'#' (an index column), track_id, artists, album_name, track_name, popularity, number, duration_ms, explicit, danceability, energynumber


Data Source:
https://www.kaggle.com/datasets/jfreyberg/spotify-chart-data
https://www.kaggle.com/datasets/maharshipandya/-spotify-tracks-dataset

### We're aiming to analyze the dataset, that can be merged on the track ID, and then answer several questions about the data.



In [6]:
df_charts_raw = pd.read_csv("charts.csv", low_memory=False);
df_dataset_raw = pd.read_csv("dataset.csv", low_memory=False);

In [28]:
df_charts_raw.head(50)

track_id               name country        date  position  \
0   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-03-17       138   
1   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-03-18       132   
2   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-03-19       128   
3   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-03-20       152   
4   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-03-21       160   
5   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-03-23       120   
6   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-06-01       172   
7   7GMeRGZax4uDKgz6wALWQw  It Was Always You      br  2015-06-03       200   
8   7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-03-17       136   
9   7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-03-18       155   
10  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-03-19       121   
11  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-03-21       192   
12  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-04-30       173   
13  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-05-06       200   
14  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-05-07       181   
15  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-05-14       190   
16  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-05-21       194   
17  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-05-28       197   
18  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-06-11       168   
19  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-06-16       189   
20  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-06-18       174   
21  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-06-25       193   
22  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-07-02       195   
23  7GMeRGZax4uDKgz6wALWQw  It Was Always You      it  2015-07-23       195   
24  7GMeRGZax4uDKgz6wALWQw  It Was Always You      ph  2015-03-19       147   
25  7GMeRGZax4uDKgz6wALWQw  It Was Always You      tw  2015-09-15       169   
26  7GMeRGZax4uDKgz6wALWQw  It Was Always You      tw  2015-09-16       167   
27  7GMeRGZax4uDKgz6wALWQw  It Was Always You      pt  2015-06-13       139   
28  7GMeRGZax4uDKgz6wALWQw  It Was Always You      pt  2015-06-16       192   
29  7GMeRGZax4uDKgz6wALWQw  It Was Always You      pt  2015-06-17       164   
30  7GMeRGZax4uDKgz6wALWQw  It Was Always You      pt  2015-06-18       164   
31  7GMeRGZax4uDKgz6wALWQw  It Was Always You      hk  2015-09-04       125   
32  7GMeRGZax4uDKgz6wALWQw  It Was Always You      hk  2015-09-05       152   
33  7GMeRGZax4uDKgz6wALWQw  It Was Always You      pe  2015-04-28        95   
34  7GMeRGZax4uDKgz6wALWQw  It Was Always You      co  2015-05-13       190   
35  7GMeRGZax4uDKgz6wALWQw  It Was Always You      cr  2016-04-15       180   
36  7GMeRGZax4uDKgz6wALWQw  It Was Always You      cr  2016-07-17       198   
37  7GMeRGZax4uDKgz6wALWQw  It Was Always You      uy  2020-03-10       176   
38  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up  global  2018-02-06       174   
39  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up  global  2018-02-07       185   
40  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up  global  2018-02-09       154   
41  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-06        63   
42  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-07        65   
43  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-08        82   
44  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-09        76   
45  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-10       120   
46  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-11       163   
47  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-12       159   
48  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      us  2018-02-13       175   
49  2Bv0MbaqYx2IrQa7Z6xBT0             T'd Up      ca  2018-02-06   

From the above view, it looks like we get a lot of row entries for each song, with different numbers of streams, and positions.
From the documentation, we can see that 'streams' is streams until x date, so we might have to use the MAX value per track_id


In [7]:
df_charts_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 626475 entries, 0 to 626474
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   track_id       626475 non-null  object
 1   name           626337 non-null  object
 2   country        626475 non-null  object
 3   date           626475 non-null  object
 4   position       626475 non-null  int64 
 5   streams        626475 non-null  object
 6   artists        626475 non-null  object
 7   artist_genres  626475 non-null  object
 8   duration       626475 non-null  int64 
 9   explicit       626475 non-null  bool  
dtypes: bool(1), int64(2), object(7)
memory usage: 43.6+ MB


### Question 1 - How many entries do we have excluding duplicate song names?

In [11]:
df_charts_raw.nunique(axis=0, dropna=True)

track_id           7188
name               6854
country              71
date               2887
position            200
streams          175019
artists            3383
artist_genres      3625
duration           6515
explicit              2
dtype: int64

Upon inspection, it looks like this dataset contains 7188 unique songs, and from those songs, we have 6854 unique names.
It looks problematic that we have 2887 unique dates. This means we might have to restrict our dataset to a shorter time period if we want to do any time series analysis...


In [24]:
print("Percent of songs with unique titles: " + str(round(6854/7188 * 100, 3)) + "%" )

Percent of songs with unique titles: 95.353%
